alignment/batch_sequential_detector.py

```diff
@@ -0,0 +1,339 @@
+                if len(segment_list) == 0: # 记录一下GPT说的胡话以便日后分析
+                    with Path('unexpected_outputs.jsonl').open('a', encoding='utf-8') as f:
+                        json.dump({'time': str(datetime.now()), 'record': record_id, 'batch': batch_index, 'input': raw_text, 'output': output_text})
+                        f.write('\n')
+                else:
+                    break
+            with filename.open('w') as f: # 只有有用的output才会被cache
+                json.dump(output_text, f)
+        else:
+            with filename.open('r') as f:
+                output_text = json.load(f)
+                segment_list = GPTBatchSequentialDetector.construct_segment_list_from_output_text(raw_text, output_text,
+                    use_identical_mapping_when_failure=True)
+            
+        return segment_list
+
+    def generate_batch(self, lines: list[str], begin_lineid: int) -> str:
+        """
+        从begin_lineid开始构造一个连续若干行的batch，使这个batch尽可能大，同时不超出self.token_limit指定的限制。
+
+        Args:
+            lines (list[str]): 源输入文件按行隔开的文本
+            begin_lineid (int): 欲开始构造的行下标，包含此行
+
+        Returns:
+            str: 构造的batch本身
+
+        """
+        # assert begin_lineid < len(lines)
+        buffer = ''
+        for lineid in range(begin_lineid, len(lines)):
+            line = lines[lineid]
+            pending_text = (buffer + '\n' if len(buffer)>0 else '') + line
+            tokens = self.encoder.encode(pending_text)
+            if len(tokens) >= self.token_limit:
+                return buffer
+            buffer = pending_text
+        if buffer:
+            return buffer
+
+    def detect(self, lines: list[str], record_id: str, **kwargs) -> list[bool]:
+        """
+        Applies the GPT-3.5 detection technique to the given lines.
+        This method first batches the lines, processes the batches, and then post-processes the results.
+
+        Args:
+            lines (list[str]): The lines to be detected.
+            record_id (int): The unique id of the record. Use to cache the output of GPT
+
+        Returns:
+            list[bool]: The detection results.
+        """
+        # processed_batches = []
+        detections = [True] * (len(lines) - 1)
+
+        new_batch_begin_lineid = 0
+        batch_id = 0
+
+        while new_batch_begin_lineid < len(lines):
+            batch = self.generate_batch(lines, new_batch_begin_lineid) # 利用已有的结果生成input_batch
+            batch_line_count = batch.count('\n') + 1
+            next_lineid = new_batch_begin_lineid + batch_line_count
+            if len(self.encoder.encode(batch)) < 20: # 结尾不能成段的噪声可能会让gpt疯狂道歉，这种情况下我们放过
+                break
+
+            # 获取成段区间表
+            segment_list = self.align_gpt_linebreak_detection_request(batch, record_id, batch_id)
+
+            for l_border, r_border in segment_list:
+                detections[new_batch_begin_lineid + l_border:new_batch_begin_lineid + r_border] = [False] * (r_border - l_border) # 每个段落的区间赋值为False
```

Comment:
If you use numpy array you can do something like 
detections[new_batch_begin_lineid + l_border:new_batch_begin_lineid + r_border] = False
(optional)